pkg/applicationserver/applicationserver.go

```diff
@@ -1055,3 +1065,33 @@ func (as *ApplicationServer) GetMQTTConfig(ctx context.Context) (*config.MQTT, e
 	}
 	return &config.MQTT, nil
 }
+
+func (as *ApplicationServer) getCachedEndDevice(ctx context.Context, ids ttnpb.EndDeviceIdentifiers) (*ttnpb.EndDevice, error) {
+	if as.deviceCache != nil {
+		dev, err := as.deviceCache.Get(ctx, ids)
+		if err != nil {
+			return dev, nil
+		}
+	}
```

Comment:
`getCachedEndDevice()` implies that it uses a cache, but that is conditional here.

Better rename this to `fetchEndDevice()` and godoc that it may use the cache?